File: src/types/alarms.ts
/** Alarm & interlock types — §13. */

export type AlarmSeverity = 'INFO' | 'WARNING' | 'ALARM' | 'TRIP'

export interface Alarm {
  id: string
  timestamp: number
  severity: AlarmSeverity
  /** Human-readable parameter name, e.g. "Rolling force". */
  parameter: string
  /** Tag the rule watches — lets the UI highlight the affected section. */
  tagName: string
  actualValue: number
  limit: number
  unit: string
  message: string
  acknowledged: boolean
  active: boolean
  /** When the alarm cleared, if it has. */
  clearedAt?: number
  /** Twin section to highlight when this alarm is active (§17 test 7). */
  section?: TwinSection
}

/** Named regions of the 3D twin that an alarm can highlight. */
export type TwinSection =
  | 'STAND'
  | 'ROLL_BITE'
  | 'ENTRY_REEL'
  | 'EXIT_REEL'
  | 'HYDRAULICS'
  | 'DRIVE'
  | 'GAUGE_ENTRY'
  | 'GAUGE_EXIT'
  | 'STRIP'

export interface AlarmRuleContext {
  value: number
  limit: number
}

export interface AlarmRule {
  id: string
  tagName: string
  parameter: string
  severity: AlarmSeverity
  unit: string
  section?: TwinSection
  /** Returns the limit that was violated, or null when the rule is satisfied. */
  evaluate: (value: number, limits: AlarmLimitLookup) => number | null
  message: (ctx: AlarmRuleContext) => string
  /** Rule only armed while the mill is in one of these states (empty = always). */
  armedStates?: string[]
}

export type AlarmLimitLookup = {
  forceWarning: number
  forceAlarm: number
  forceTrip: number
  thicknessToleranceUm: number
  hydraulicPressureMin: number
  entryTensionMin: number
  entryTensionMax: number
  exitTensionMin: number
  exitTensionMax: number
  currentMax: number
  staleAfterMs: number
}

export interface InterlockNode {
  id: string
  label: string
  ok: boolean
  /** Why it is not OK. Never show a bare "NOT READY" (§13.2). */
  reason: string
}

export interface InterlockChain {
  nodes: InterlockNode[]
  millReady: boolean
  /** First failing node, used for the "Reason: …" line. */
  blockingReason: string | null
}

File: dotnet/src/Crm04.Domain/Types/Alarms.cs
namespace Crm04.Domain.Types;

/// <summary>
/// Port of <c>src/types/alarms.ts</c> - the latched alarm as the UI sees it.
///
/// The pure rule evaluation produces an <see cref="AlarmCondition"/>; latching, acknowledgement
/// and clear-time bookkeeping turn that into an <see cref="Alarm"/>. Keeping them as two types
/// is what lets the rules stay a pure function of <see cref="MachineState"/>.
/// </summary>
/// <param name="Id">Stable rule id, e.g. <c>HIGH_ROLLING_FORCE</c>.</param>
/// <param name="Parameter">Human-readable parameter name, e.g. "Rolling force".</param>
/// <param name="TagName">Tag the rule watches - lets the UI highlight the affected section.</param>
/// <param name="Section">Twin section to highlight while this alarm is active (§17 test 7).</param>
/// <param name="ClearedAt">When the alarm cleared, if it has.</param>
public sealed record Alarm(
    string Id,
    long Timestamp,
    AlarmSeverity Severity,
    string Parameter,
    string TagName,
    double ActualValue,
    double Limit,
    string Unit,
    string Message,
    bool Acknowledged,
    bool Active,
    long? ClearedAt = null,
    TwinSection? Section = null);

/// <summary>
/// One violated rule at one instant. Produced by the pure alarm engine; carries no timestamp,
/// no acknowledgement and no active flag, because none of those are properties of the rule -
/// they belong to the latching layer above it.
/// </summary>
public sealed record AlarmCondition(
    string Id,
    AlarmSeverity Severity,
    string Parameter,
    string TagName,
    double ActualValue,
    double Limit,
    string Unit,
    string Message,
    TwinSection? Section = null);

/// <summary>
/// One link in the interlock chain. <paramref name="Reason"/> is never empty when
/// <paramref name="Ok"/> is false - §13.2 forbids showing a bare "NOT READY".
/// </summary>
public sealed record InterlockNode(string Id, string Label, bool Ok, string Reason);

/// <param name="BlockingReason">First failing node, used for the "Reason: …" line. Null when ready.</param>
public sealed record InterlockChain(
    IReadOnlyList<InterlockNode> Nodes,
    bool MillReady,
    string? BlockingReason)
{
    public static readonly InterlockChain Empty = new([], false, "NO DATA");
}
